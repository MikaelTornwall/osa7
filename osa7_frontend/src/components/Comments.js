import React from 'react'
import { addComment } from '../reducers/blogReducer'
import { notify } from '../reducers/notificationReducer'
import { connect } from 'react-redux'
import { Comment, Header, Form, Button, Icon } from 'semantic-ui-react'

class Comments extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      comment: ''
    }
  }

  handleChange = (event) => {
    console.log(event.target.value)
    this.setState({
      comment: event.target.value
    })
  }

  render() {
    const { blog } = this.props
    const comments = blog.comment ? blog.comment.map(comment => comment) : []

    const newComment = () => {
      return async (event) => {
        event.preventDefault()
        const commentObject = {
          comment: this.state.comment,
          blogId: blog.id
        }
        this.props.notify(`Comment "${commentObject.comment}" added to blog "${blog.title}".`)
        this.setState({
          comment: ''
        })

        await this.props.addComment(commentObject)
      }
    }

    return (
      <Comment.Group size='large'>
        <Header as='h3' dividing>Comments ({comments.length})</Header>
        {comments.map(comment =>
          <Comment key={comment._id}>
            <Icon name='comment outline' />
            <Comment.Metadata>{comment.date}</Comment.Metadata>
            <Comment.Content>
              {comment.comment}
            </Comment.Content>
          </Comment>
        )}
        <br />
        <div>
          <Form reply onSubmit={newComment()}>
            <Form.TextArea name="comment" value={this.state.comment} onChange={this.handleChange} />
            <Button content='Comment' type='submit' icon='edit' color='teal' />
          </Form>
        </div>
      </Comment.Group>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    blog: state.blog
  }
}

const mapDispatchToProps = {
  addComment,
  notify
}

const ConnectedComments = connect(
  mapStateToProps,
  mapDispatchToProps
)(Comments)

export default ConnectedComments
