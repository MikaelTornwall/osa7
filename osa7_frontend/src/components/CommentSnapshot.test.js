import React from 'react'
import { Comments } from './Comments'
import renderer from 'react-test-renderer'

describe('<Comments />', () => {
  it('renders correctly', () => {
    const blog = {
      id: "5b392c65fa9613b090a7884b",
      likes: 5,
      title: "TechCrunch",
      author: "Michael Arrington & Keith Teare",
      url: "https://techcrunch.com/",
      user: {
        _id: "5b2fc1f7581f5d6dde0cabc2",
        username: "mikael"
      },
      comment: [
        {
          _id: "5b3990fd462d3cbd1e33d8d2",
          comment: "Very good!",
          date: "Sun Jul 01 2018 20:42:05 GMT-0600 (CST)"
        }
      ]
    }

    const tree = renderer
      .create(<Comments blog={blog} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
