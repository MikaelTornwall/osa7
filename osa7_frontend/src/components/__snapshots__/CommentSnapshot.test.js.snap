// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Comments /> renders correctly 1`] = `
<div
  className="ui large comments"
>
  <h3
    className="ui dividing header"
  >
    Comments 
    <span
      className="comment-header"
    >
      (
      1
      )
    </span>
  </h3>
  <div
    className="comment"
  >
    <i
      aria-hidden="true"
      className="comment outline icon"
    />
    <div
      className="metadata"
    >
      Sun Jul 01 2018 20:42:05 GMT-0600 (CST)
    </div>
    <div
      className="content"
    >
      <div
        className="comment"
      >
        Very good!
      </div>
    </div>
  </div>
  <br />
  <div>
    <form
      action={undefined}
      className="ui reply form"
      onSubmit={[Function]}
    >
      <div
        className="field"
      >
        <textarea
          content={undefined}
          disabled={undefined}
          name="comment"
          onChange={[Function]}
          onInput={[Function]}
          required={undefined}
          rows={3}
          style={
            Object {
              "resize": "",
            }
          }
          type={undefined}
          value=""
        />
      </div>
      <button
        className="ui teal button"
        disabled={undefined}
        onClick={[Function]}
        role="button"
        tabIndex={undefined}
        type="submit"
      >
        <i
          aria-hidden="true"
          className="edit icon"
        />
        Comment
      </button>
    </form>
  </div>
</div>
`;
